File: hospital_server.py
import hashlib
import json
import socket
from pathlib import Path


HOST = "127.0.0.1"
USC_ID_SUFFIX = 818

AUTH_UDP_PORT = 21000 + USC_ID_SUFFIX  # 21818
PRESC_UDP_PORT = 22000 + USC_ID_SUFFIX  # 22818
APPT_UDP_PORT = 23000 + USC_ID_SUFFIX  # 23818

HOSP_UDP_PORT = 25000 + USC_ID_SUFFIX  # 25818
HOSP_TCP_PORT = 26000 + USC_ID_SUFFIX  # 26818


def sha256_hash(text: str) -> str:
    text = text.strip()
    return hashlib.sha256(text.encode('utf-8')).hexdigest()


def hash_suffix(sha256_hex: str) -> str:
    return sha256_hex[-5:]


def recv_json_line(conn: socket.socket) -> dict | None:
    buf = b""
    while b"\n" not in buf:
        chunk = conn.recv(4096)
        if not chunk:
            return None
        buf += chunk
        if len(buf) > 1024 * 1024:
            return None
    line, _rest = buf.split(b"\n", 1)
    try:
        return json.loads(line.decode("utf-8"))
    except Exception:
        return None


def send_json_line(conn: socket.socket, obj: dict) -> None:
    conn.sendall(json.dumps(obj).encode("utf-8") + b"\n")


def udp_rpc(udp_sock: socket.socket, dest_port: int, req: dict) -> dict | None:
    udp_sock.sendto(json.dumps(req).encode("utf-8"), (HOST, dest_port))
    udp_sock.settimeout(3.0)
    try:
        data, _addr = udp_sock.recvfrom(65535)
    except TimeoutError:
        return None
    finally:
        udp_sock.settimeout(None)
    try:
        return json.loads(data.decode("utf-8"))
    except Exception:
        return None


def load_doctors_and_treatments(path: Path) -> tuple[dict[str, str], dict[str, str]]:
    """
    hospital.txt:
      [Doctors]
      <doctor_name> <hashed_doctor_name>
      ...
      [Treatments]
      <illness> <treatment>
      ...
    Returns:
      doctor_hash -> doctor_name
      illness -> treatment
    """
    doctor_hash_to_name: dict[str, str] = {}
    illness_to_treatment: dict[str, str] = {}
    if not path.exists():
        return doctor_hash_to_name, illness_to_treatment

    section = None
    for raw in path.read_text(encoding="utf-8").splitlines():
        s = raw.strip()
        if not s:
            continue
        if s == "[Doctors]":
            section = "doctors"
            continue
        if s == "[Treatments]":
            section = "treatments"
            continue
        parts = s.split()
        if section == "doctors" and len(parts) >= 2:
            name, h = parts[0], parts[1]
            doctor_hash_to_name[h] = name
        elif section == "treatments" and len(parts) >= 2:
            illness, treatment = parts[0], parts[1]
            illness_to_treatment[illness] = treatment

    return doctor_hash_to_name, illness_to_treatment


def main() -> None:
    hospital_txt = Path("hospital.txt")

    udp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    udp_sock.bind((HOST, HOSP_UDP_PORT))

    tcp_listen = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    tcp_listen.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    tcp_listen.bind((HOST, HOSP_TCP_PORT))
    tcp_listen.listen(10)

    print(f"Hospital Server is up and running using UDP on port\n{HOSP_UDP_PORT}.")

    while True:
        conn, _addr = tcp_listen.accept()
        try:
            msg = recv_json_line(conn)
            if msg is None:
                continue

            mtype = msg.get("type")

            if mtype == "auth":
                u_hash = str(msg.get("u_hash", ""))
                p_hash = str(msg.get("p_hash", ""))
                suffix = hash_suffix(u_hash)

                print(f"Hospital Server received an authentication request from a\nuser with hash suffix {suffix}.")
                print("Hospital Server has sent an authentication request to the\nAuthentication Server.")

                resp = udp_rpc(udp_sock, AUTH_UDP_PORT, {"type": "auth_req", "u_hash": u_hash, "p_hash": p_hash})
                print(
                    "Hospital server has received the response from the\n"
                    f"authentication server using UDP over port {HOSP_UDP_PORT}."
                )

                ok = bool(resp and resp.get("ok"))
                if not ok:
                    send_json_line(conn, {"type": "auth_result", "ok": False})
                    continue

                print(f"User with a hash suffix {suffix} has been granted\naccess to the system. Determining the access of the user.")

                doctor_hash_to_name, _treat = load_doctors_and_treatments(hospital_txt)
                is_doctor = u_hash in doctor_hash_to_name
                if is_doctor:
                    print(f"User with hash suffix {suffix} will be granted doctor\naccess.")
                    role = "doctor"
                    doctor_name = doctor_hash_to_name[u_hash]
                else:
                    print(f"User with hash {suffix} will be granted patient\naccess.")
                    role = "patient"
                    doctor_name = None

                send_json_line(conn, {"type": "auth_result", "ok": True, "role": role, "doctor_name": doctor_name})
                print(
                    "Hospital Server has sent the response from\n"
                    f"Authentication Server to the client using TCP over port\n{HOSP_TCP_PORT}."
                )

            elif mtype == "lookup_doctors":
                u_hash = str(msg.get("u_hash", ""))
                suffix = hash_suffix(u_hash)
                print(
                    "Hospital Server received a lookup request from a user\n"
                    f"with a hash suffix {suffix} over port {HOSP_TCP_PORT}."
                )
                print("Hospital Server sent the doctor lookup request to the\nAppointment server.")
                resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "doctor_list_req"})
                print(
                    "Hospital Server has received the response from\n"
                    f"Appointment Server using UDP over port {HOSP_UDP_PORT}."
                )
                doctors = resp.get("doctors", []) if resp else []
                send_json_line(conn, {"type": "lookup_doctors_resp", "doctors": doctors})
                print("Hospital Server has sent the doctor lookup to the client.")

            elif mtype == "lookup_availability":
                u_hash = str(msg.get("u_hash", ""))
                doctor = str(msg.get("doctor", ""))
                suffix = hash_suffix(u_hash)
                print(
                    "Hospital Server has received a lookup request from a\n"
                    f"user with hash suffix {suffix} to lookup\n"
                    f"{doctor} availability using TCP over port\n{HOSP_TCP_PORT}."
                )
                print("Hospital Server sent the doctor lookup request to the\nAppointment server.")
                resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "doctor_avail_req", "doctor": doctor})
                print(
                    "Hospital Server has received the response from\n"
                    f"Appointment Server using UDP over port {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "lookup_availability_resp", "doctor": doctor, "avail": (resp or {}).get("avail", [])})
                print("The Hospital Server has sent the response to the client.")

            elif mtype == "schedule":
                u_hash = str(msg.get("u_hash", ""))
                doctor = str(msg.get("doctor", ""))
                time_block = str(msg.get("time", ""))
                illness = str(msg.get("illness", ""))
                suffix = hash_suffix(u_hash)
                print(
                    "Hospital Server has received a schedule request from a\n"
                    f"user with hash suffix: {suffix} to book an\n"
                    f"appointment using TCP over port {HOSP_TCP_PORT}."
                )
                print("Hospital Server has sent the schedule request to the\nappointment server.")
                resp = udp_rpc(
                    udp_sock,
                    APPT_UDP_PORT,
                    {"type": "schedule_req", "doctor": doctor, "time": time_block, "patient_hash": u_hash, "illness": illness},
                )
                print(
                    "Hospital Server has received the response from\n"
                    f"Appointment Server using UDP over {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "schedule_resp", **(resp or {"ok": False, "doctor": doctor, "time": time_block, "other_avail": []})})
                print("The hospital server has sent the response to the client.")

            elif mtype == "cancel":
                u_hash = str(msg.get("u_hash", ""))
                suffix = hash_suffix(u_hash)
                print(
                    "Hospital Server has received a cancel request from user\n"
                    f"with hash suffix: {suffix} to cancel their appointment\n"
                    f"using TCP over port {HOSP_TCP_PORT}."
                )
                print("The hospital server has sent the cancel request to the\nappointment server.")
                resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "cancel_req", "patient_hash": u_hash})
                print(
                    "Hospital Server has received the response from\n"
                    f"Appointment Server using UDP over port {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "cancel_resp", **(resp or {"ok": False})})
                print("The hospital server has sent the response to the client.")

            elif mtype == "view_appointment":
                u_hash = str(msg.get("u_hash", ""))
                suffix = hash_suffix(u_hash)
                print(
                    "Hospital server has received a view appointment request\n"
                    f"from a user with hash suffix {suffix} to view their\n"
                    f"appointment details using TCP over port {HOSP_TCP_PORT}."
                )
                print("Hospital Server has sent the view appointments request\nto the Appointment Server.")
                resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "view_appt_req", "patient_hash": u_hash})
                print(
                    "Hospital Server has received the response from the\n"
                    f"appointment server using UDP over port {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "view_appointment_resp", **(resp or {"ok": False})})
                print("The hospital server has sent the response to the client.")

            elif mtype == "view_appointments":
                doctor = str(msg.get("doctor", ""))
                print(
                    "Hospital Server has received a view appointments\n"
                    f"request from {doctor} to view their schedule\n"
                    f"details using TCP over port {HOSP_TCP_PORT}."
                )
                print("The hospital server has sent the view appointments\nrequest to the Appointment Server.")
                resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "view_appointments_req", "doctor": doctor})
                print(
                    "Hospital server has received the response from the\n"
                    f"Appointment server using UDP over port {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "view_appointments_resp", **(resp or {"doctor": doctor, "times": []})})
                print("The hospital server has sent the response to the client.")

            elif mtype == "prescribe":
                doctor = str(msg.get("doctor", ""))
                patient_hash = str(msg.get("patient_hash", ""))
                frequency = str(msg.get("frequency", ""))
                print(
                    "Hospital Server has received a prescription request from\n"
                    f"{doctor} for a user with hash suffix\n"
                    f"{hash_suffix(patient_hash)} using TCP over port {HOSP_TCP_PORT}."
                )
                print(
                    "Hospital Server has sent a request to fetch patients with\n"
                    f"hash suffix {hash_suffix(patient_hash)} illness information to the\nAppointment Server."
                )
                illness_resp = udp_rpc(udp_sock, APPT_UDP_PORT, {"type": "fetch_illness_req", "doctor": doctor, "patient_hash": patient_hash})
                print(
                    "Hospital Server has received the illness response from\n"
                    f"the Appointment server using UDP over port {HOSP_UDP_PORT}."
                )
                illness = (illness_resp or {}).get("illness")
                doctor_hash_to_name, illness_to_treatment = load_doctors_and_treatments(hospital_txt)
                if not illness:
                    send_json_line(conn, {"type": "prescribe_resp", "ok": False})
                    continue
                print(f"Acquiring treatment for {illness} from the database.")
                treatment = illness_to_treatment.get(str(illness), "None")
                print(f"Hospital server has sent the prescription request to the\nprescription server to prescribe {treatment}.")
                presc_resp = udp_rpc(
                    udp_sock,
                    PRESC_UDP_PORT,
                    {"type": "prescribe_req", "doctor": doctor, "patient_hash": patient_hash, "treatment": treatment, "frequency": frequency},
                )
                print(
                    "Hospital server has received the response from the\n"
                    f"prescription server using UDP over port {HOSP_UDP_PORT}"
                )
                send_json_line(conn, {"type": "prescribe_resp", "ok": bool(presc_resp and presc_resp.get("ok")), "treatment": treatment})
                print("The hospital server has sent the response to the client.")

            elif mtype == "view_prescription":
                # Caller can be patient or doctor; we just forward to prescription server.
                requester_role = str(msg.get("role", "patient"))
                doctor = msg.get("doctor")
                patient_hash = str(msg.get("patient_hash", ""))
                if requester_role == "patient":
                    print(
                        "Hospital Server has received a prescription request from\n"
                        f"a patient with hash suffix {hash_suffix(patient_hash)} to view their\n"
                        f"prescription details using TCP over port {HOSP_TCP_PORT}."
                    )
                else:
                    print(
                        "Hospital Server has received a prescription request from\n"
                        f"{doctor} to view a patient with hash suffix\n"
                        f"{hash_suffix(patient_hash)} prescription details using TCP over port\n{HOSP_TCP_PORT}."
                    )
                print("Hospital Server has sent the prescription request to the\nPrescription Server.")
                resp = udp_rpc(udp_sock, PRESC_UDP_PORT, {"type": "view_prescription_req", "patient_hash": patient_hash})
                print(
                    "Hospital server has received the response from the\n"
                    f"prescription server using UDP over port {HOSP_UDP_PORT}."
                )
                send_json_line(conn, {"type": "view_prescription_resp", **(resp or {"ok": False})})
                print("Hospital server has sent the response to the client.")

        finally:
            try:
                conn.close()
            except Exception:
                pass


if __name__ == "__main__":
    main()

File: hospital_server.cpp
#include "constants.h"
#include "crypto_utils.h"
#include "file_utils.h"
#include "net_utils.h"
#include "text_proto.h"

#include <arpa/inet.h>
#include <sys/socket.h>
#include <unistd.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

static std::vector<std::string> split_ws(const std::string &s) {
    std::vector<std::string> raw = split_char(s, ' ');
    std::vector<std::string> out;
    for (size_t i = 0; i < raw.size(); ++i) {
        if (!raw[i].empty()) {
            out.push_back(raw[i]);
        }
    }
    return out;
}

static void load_hospital_data(
    const std::string &path,
    std::map<std::string, std::string> &doctor_hash_to_name,
    std::map<std::string, std::string> &illness_to_treatment) {
    doctor_hash_to_name.clear();
    illness_to_treatment.clear();
    std::vector<std::string> lines;
    if (!read_lines(path, lines)) {
        return;
    }
    std::string section;
    for (size_t i = 0; i < lines.size(); ++i) {
        std::string s = trim_copy(lines[i]);
        if (s.empty()) {
            continue;
        }
        if (s == "[Doctors]") {
            section = "doctors";
            continue;
        }
        if (s == "[Treatments]") {
            section = "treatments";
            continue;
        }
        std::vector<std::string> p = split_ws(s);
        if (section == "doctors" && p.size() >= 2) {
            doctor_hash_to_name[p[1]] = p[0];
        } else if (section == "treatments" && p.size() >= 2) {
            illness_to_treatment[p[0]] = p[1];
        }
    }
}

static bool udp_rpc(int udp_fd, int dest_port, const ProtoMessage &req, ProtoMessage &resp) {
    std::string encoded = proto_serialize(req);
    if (!udp_send_to(udp_fd, HOST, dest_port, encoded)) {
        return false;
    }
    std::string payload;
    if (!udp_recv_with_timeout(udp_fd, 3000, payload)) {
        return false;
    }
    return proto_parse(payload, resp);
}

static std::string getf(const ProtoMessage &m, const std::string &key, const std::string &fallback = "") {
    std::map<std::string, std::string>::const_iterator it = m.fields.find(key);
    if (it == m.fields.end()) {
        return fallback;
    }
    return it->second;
}

static bool is_true(const std::string &v) {
    return v == "1" || v == "true" || v == "True";
}

int main() {
    int udp_fd = create_udp_bound_socket(HOST, HOSP_UDP_PORT);
    if (udp_fd < 0) {
        return 1;
    }
    int listener = create_tcp_listener(HOST, HOSP_TCP_PORT, 10);
    if (listener < 0) {
        close(udp_fd);
        return 1;
    }

    std::cout << "Hospital Server is up and running using UDP on port\n" << HOSP_UDP_PORT << "." << std::endl;

    while (true) {
        int conn = accept(listener, NULL, NULL);
        if (conn < 0) {
            continue;
        }

        std::string line;
        if (!recv_line_tcp(conn, line)) {
            close(conn);
            continue;
        }

        ProtoMessage req;
        if (!proto_parse(line, req)) {
            close(conn);
            continue;
        }

        if (req.type == "auth") {
            std::string u_hash = getf(req, "u_hash");
            std::string p_hash = getf(req, "p_hash");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital Server received an authentication request from a\nuser with hash suffix " << suffix << "." << std::endl;
            std::cout << "Hospital Server has sent an authentication request to the\nAuthentication Server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "auth_req";
            a_req.fields["u_hash"] = u_hash;
            a_req.fields["p_hash"] = p_hash;
            ProtoMessage a_resp;
            bool ok_rpc = udp_rpc(udp_fd, AUTH_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital server has received the response from the\nauthentication server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "auth_result";
            bool ok = ok_rpc && is_true(getf(a_resp, "ok", "0"));
            if (!ok) {
                out.fields["ok"] = "0";
                send_all(conn, proto_serialize(out) + "\n");
                close(conn);
                continue;
            }

            std::cout << "User with a hash suffix " << suffix << " has been granted\naccess to the system. Determining the access of the user." << std::endl;
            std::map<std::string, std::string> doctor_hash_to_name;
            std::map<std::string, std::string> illness_to_treatment;
            load_hospital_data("hospital.txt", doctor_hash_to_name, illness_to_treatment);
            std::map<std::string, std::string>::iterator dit = doctor_hash_to_name.find(u_hash);
            if (dit != doctor_hash_to_name.end()) {
                std::cout << "User with hash suffix " << suffix << " will be granted doctor\naccess." << std::endl;
                out.fields["ok"] = "1";
                out.fields["role"] = "doctor";
                out.fields["doctor_name"] = dit->second;
            } else {
                std::cout << "User with hash " << suffix << " will be granted patient\naccess." << std::endl;
                out.fields["ok"] = "1";
                out.fields["role"] = "patient";
                out.fields["doctor_name"] = "";
            }
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "Hospital Server has sent the response from\nAuthentication Server to the client using TCP over port\n"
                      << HOSP_TCP_PORT << "." << std::endl;
        } else if (req.type == "lookup_doctors") {
            std::string u_hash = getf(req, "u_hash");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital Server received a lookup request from a user\nwith a hash suffix " << suffix << " over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "Hospital Server sent the doctor lookup request to the\nAppointment server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "doctor_list_req";
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital Server has received the response from\nAppointment Server using UDP over port " << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "lookup_doctors_resp";
            out.fields["doctors"] = getf(a_resp, "doctors");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "Hospital Server has sent the doctor lookup to the client." << std::endl;
        } else if (req.type == "lookup_availability") {
            std::string u_hash = getf(req, "u_hash");
            std::string doctor = getf(req, "doctor");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital Server has received a lookup request from a\nuser with hash suffix " << suffix
                      << " to lookup\n" << doctor << " availability using TCP over port\n" << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "Hospital Server sent the doctor lookup request to the\nAppointment server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "doctor_avail_req";
            a_req.fields["doctor"] = doctor;
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital Server has received the response from\nAppointment Server using UDP over port " << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "lookup_availability_resp";
            out.fields["doctor"] = doctor;
            out.fields["avail"] = getf(a_resp, "avail");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The Hospital Server has sent the response to the client." << std::endl;
        } else if (req.type == "schedule") {
            std::string u_hash = getf(req, "u_hash");
            std::string doctor = getf(req, "doctor");
            std::string time = getf(req, "time");
            std::string illness = getf(req, "illness");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital Server has received a schedule request from a\nuser with hash suffix: " << suffix
                      << " to book an\nappointment using TCP over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "Hospital Server has sent the schedule request to the\nappointment server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "schedule_req";
            a_req.fields["doctor"] = doctor;
            a_req.fields["time"] = time;
            a_req.fields["patient_hash"] = u_hash;
            a_req.fields["illness"] = illness;
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital Server has received the response from\nAppointment Server using UDP over "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "schedule_resp";
            out.fields["ok"] = getf(a_resp, "ok", "0");
            out.fields["doctor"] = doctor;
            out.fields["time"] = time;
            out.fields["other_avail"] = getf(a_resp, "other_avail");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The hospital server has sent the response to the client." << std::endl;
        } else if (req.type == "cancel") {
            std::string u_hash = getf(req, "u_hash");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital Server has received a cancel request from user\nwith hash suffix: " << suffix
                      << " to cancel their appointment\nusing TCP over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "The hospital server has sent the cancel request to the\nappointment server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "cancel_req";
            a_req.fields["patient_hash"] = u_hash;
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital Server has received the response from\nAppointment Server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "cancel_resp";
            out.fields["ok"] = getf(a_resp, "ok", "0");
            out.fields["doctor"] = getf(a_resp, "doctor");
            out.fields["time"] = getf(a_resp, "time");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The hospital server has sent the response to the client." << std::endl;
        } else if (req.type == "view_appointment") {
            std::string u_hash = getf(req, "u_hash");
            std::string suffix = hash_suffix5(u_hash);
            std::cout << "Hospital server has received a view appointment request\nfrom a user with hash suffix " << suffix
                      << " to view their\nappointment details using TCP over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "Hospital Server has sent the view appointments request\nto the Appointment Server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "view_appt_req";
            a_req.fields["patient_hash"] = u_hash;
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital Server has received the response from the\nappointment server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "view_appointment_resp";
            out.fields["ok"] = getf(a_resp, "ok", "0");
            out.fields["doctor"] = getf(a_resp, "doctor");
            out.fields["time"] = getf(a_resp, "time");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The hospital server has sent the response to the client." << std::endl;
        } else if (req.type == "view_appointments") {
            std::string doctor = getf(req, "doctor");
            std::cout << "Hospital Server has received a view appointments\nrequest from " << doctor
                      << " to view their schedule\ndetails using TCP over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "The hospital server has sent the view appointments\nrequest to the Appointment Server." << std::endl;
            ProtoMessage a_req;
            a_req.type = "view_appointments_req";
            a_req.fields["doctor"] = doctor;
            ProtoMessage a_resp;
            udp_rpc(udp_fd, APPT_UDP_PORT, a_req, a_resp);
            std::cout << "Hospital server has received the response from the\nAppointment server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "view_appointments_resp";
            out.fields["doctor"] = doctor;
            out.fields["times"] = getf(a_resp, "times");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The hospital server has sent the response to the client." << std::endl;
        } else if (req.type == "prescribe") {
            std::string doctor = getf(req, "doctor");
            std::string patient_hash = getf(req, "patient_hash");
            std::string frequency = getf(req, "frequency");
            std::cout << "Hospital Server has received a prescription request from\n" << doctor
                      << " for a user with hash suffix\n" << hash_suffix5(patient_hash)
                      << " using TCP over port " << HOSP_TCP_PORT << "." << std::endl;
            std::cout << "Hospital Server has sent a request to fetch patients with\nhash suffix " << hash_suffix5(patient_hash)
                      << " illness information to the\nAppointment Server." << std::endl;

            ProtoMessage fetch_req;
            fetch_req.type = "fetch_illness_req";
            fetch_req.fields["doctor"] = doctor;
            fetch_req.fields["patient_hash"] = patient_hash;
            ProtoMessage fetch_resp;
            bool got_fetch = udp_rpc(udp_fd, APPT_UDP_PORT, fetch_req, fetch_resp);
            std::cout << "Hospital Server has received the illness response from\nthe Appointment server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;

            std::string illness = got_fetch ? getf(fetch_resp, "illness") : "";
            std::map<std::string, std::string> doctor_hash_to_name;
            std::map<std::string, std::string> illness_to_treatment;
            load_hospital_data("hospital.txt", doctor_hash_to_name, illness_to_treatment);
            std::string treatment = "None";

            ProtoMessage out;
            out.type = "prescribe_resp";
            if (illness.empty()) {
                out.fields["ok"] = "0";
                send_all(conn, proto_serialize(out) + "\n");
                close(conn);
                continue;
            }

            std::cout << "Acquiring treatment for " << illness << " from the database." << std::endl;
            if (illness_to_treatment.find(illness) != illness_to_treatment.end()) {
                treatment = illness_to_treatment[illness];
            }
            std::cout << "Hospital server has sent the prescription request to the\nprescription server to prescribe "
                      << treatment << "." << std::endl;
            ProtoMessage p_req;
            p_req.type = "prescribe_req";
            p_req.fields["doctor"] = doctor;
            p_req.fields["patient_hash"] = patient_hash;
            p_req.fields["treatment"] = treatment;
            p_req.fields["frequency"] = frequency;
            ProtoMessage p_resp;
            bool got_p = udp_rpc(udp_fd, PRESC_UDP_PORT, p_req, p_resp);
            std::cout << "Hospital server has received the response from the\nprescription server using UDP over port "
                      << HOSP_UDP_PORT << std::endl;
            out.fields["ok"] = (got_p && is_true(getf(p_resp, "ok", "0"))) ? "1" : "0";
            out.fields["treatment"] = treatment;
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "The hospital server has sent the response to the client." << std::endl;
        } else if (req.type == "view_prescription") {
            std::string role = getf(req, "role", "patient");
            std::string doctor = getf(req, "doctor");
            std::string patient_hash = getf(req, "patient_hash");
            if (role == "patient") {
                std::cout << "Hospital Server has received a prescription request from\na patient with hash suffix "
                          << hash_suffix5(patient_hash) << " to view their\nprescription details using TCP over port "
                          << HOSP_TCP_PORT << "." << std::endl;
            } else {
                std::cout << "Hospital Server has received a prescription request from\n" << doctor
                          << " to view a patient with hash suffix\n" << hash_suffix5(patient_hash)
                          << " prescription details using TCP over port\n" << HOSP_TCP_PORT << "." << std::endl;
            }
            std::cout << "Hospital Server has sent the prescription request to the\nPrescription Server." << std::endl;
            ProtoMessage p_req;
            p_req.type = "view_prescription_req";
            p_req.fields["patient_hash"] = patient_hash;
            ProtoMessage p_resp;
            udp_rpc(udp_fd, PRESC_UDP_PORT, p_req, p_resp);
            std::cout << "Hospital server has received the response from the\nprescription server using UDP over port "
                      << HOSP_UDP_PORT << "." << std::endl;
            ProtoMessage out;
            out.type = "view_prescription_resp";
            out.fields["ok"] = getf(p_resp, "ok", "0");
            out.fields["doctor"] = getf(p_resp, "doctor");
            out.fields["treatment"] = getf(p_resp, "treatment");
            out.fields["frequency"] = getf(p_resp, "frequency");
            send_all(conn, proto_serialize(out) + "\n");
            std::cout << "Hospital server has sent the response to the client." << std::endl;
        }

        close(conn);
    }

    close(listener);
    close(udp_fd);
    return 0;
}
